File: DevDocs/backend/tests/test-document-cache.js
/**
 * Document Cache Service Tests
 * 
 * Tests for the document caching functionality.
 */

const path = require('path');
const fs = require('fs');
const { expect } = require('chai');
const crypto = require('crypto');

// Import the CachedSecurityExtractor
const CachedSecurityExtractor = require('../enhanced_processing/security_extractor_cached');

// Test configuration
const TEST_CACHE_DIR = path.resolve(__dirname, './test-cache');
const TEST_PDF_PATH = path.resolve(__dirname, './samples/sample-portfolio.pdf');
const TEST_TENANT_ID = 'test-tenant-' + Date.now();

// Create test directory
if (!fs.existsSync(TEST_CACHE_DIR)) {
  fs.mkdirSync(TEST_CACHE_DIR, { recursive: true });
}

// Prepare a test PDF if one doesn't exist
if (!fs.existsSync(TEST_PDF_PATH)) {
  const testPdfDir = path.dirname(TEST_PDF_PATH);
  if (!fs.existsSync(testPdfDir)) {
    fs.mkdirSync(testPdfDir, { recursive: true });
  }
  
  // Create a simple PDF-like file for testing
  const dummyContent = Buffer.from('%PDF-1.5\nTest PDF content\n%%EOF');
  fs.writeFileSync(TEST_PDF_PATH, dummyContent);
}

// Mock the extract_from_pdf method for testing
class TestSecurityExtractor extends CachedSecurityExtractor {
  constructor(options) {
    super(options);
    this.extractCallCount = 0;
  }
  
  async extract_from_pdf(pdfPath) {
    this.extractCallCount++;
    
    // Return mock data
    return {
      securities: [
        { isin: 'US0378331005', name: 'Apple Inc.', value: 1000 },
        { isin: 'US5949181045', name: 'Microsoft Corp.', value: 2000 }
      ],
      timestamp: new Date().toISOString(),
      document: path.basename(pdfPath)
    };
  }
}

// Test suite
describe('Document Cache Service', function() {
  this.timeout(10000); // Set timeout to 10 seconds
  
  let extractor;
  
  before(() => {
    // Set up the test extractor
    extractor = new TestSecurityExtractor({
      debug: true,
      cacheDir: TEST_CACHE_DIR,
      cacheTtl: 60, // 1 minute
      useTenantIsolation: true
    });
  });
  
  after(() => {
    // Clean up test directory
    if (fs.existsSync(TEST_CACHE_DIR)) {
      fs.rmSync(TEST_CACHE_DIR, { recursive: true, force: true });
    }
  });
  
  it('should generate consistent document fingerprint', () => {
    const fingerprint1 = extractor.generateDocumentFingerprint(TEST_PDF_PATH);
    const fingerprint2 = extractor.generateDocumentFingerprint(TEST_PDF_PATH);
    
    expect(fingerprint1).to.be.a('string');
    expect(fingerprint1.length).to.be.greaterThan(0);
    expect(fingerprint1).to.equal(fingerprint2);
  });
  
  it('should include metadata in document fingerprint', () => {
    const metadata = { documentType: 'portfolio', userId: '12345' };
    
    const fingerprintWithoutMetadata = extractor.generateDocumentFingerprint(TEST_PDF_PATH);
    const fingerprintWithMetadata = extractor.generateDocumentFingerprint(TEST_PDF_PATH, metadata);
    
    expect(fingerprintWithMetadata).to.not.equal(fingerprintWithoutMetadata);
  });
  
  it('should cache document processing results', async () => {
    const result1 = await extractor.extract_from_pdf_cached(TEST_PDF_PATH, {
      tenantId: TEST_TENANT_ID
    });
    
    expect(result1).to.be.an('object');
    expect(result1.securities).to.be.an('array');
    expect(result1.securities.length).to.equal(2);
    expect(extractor.extractCallCount).to.equal(1);
    
    // Extract again to test cache
    const result2 = await extractor.extract_from_pdf_cached(TEST_PDF_PATH, {
      tenantId: TEST_TENANT_ID
    });
    
    expect(result2).to.deep.equal(result1);
    expect(extractor.extractCallCount).to.equal(1); // Should not increase
  });
  
  it('should honor the forceRefresh flag', async () => {
    // Extract with force refresh
    const result = await extractor.extract_from_pdf_cached(TEST_PDF_PATH, {
      tenantId: TEST_TENANT_ID,
      forceRefresh: true
    });
    
    expect(result).to.be.an('object');
    expect(extractor.extractCallCount).to.equal(2); // Should increase
  });
  
  it('should respect tenant isolation', async () => {
    const tenant1 = 'tenant1';
    const tenant2 = 'tenant2';
    
    // Extract for tenant1
    await extractor.extract_from_pdf_cached(TEST_PDF_PATH, {
      tenantId: tenant1
    });
    
    // Extract for tenant2
    await extractor.extract_from_pdf_cached(TEST_PDF_PATH, {
      tenantId: tenant2
    });
    
    expect(extractor.extractCallCount).to.equal(4); // Should increase twice
    
    // Check cache directories were created
    expect(fs.existsSync(path.join(TEST_CACHE_DIR, tenant1))).to.be.true;
    expect(fs.existsSync(path.join(TEST_CACHE_DIR, tenant2))).to.be.true;
  });
  
  it('should invalidate cache entries', async () => {
    const tenantId = 'invalidation-test';
    
    // Extract and cache
    await extractor.extract_from_pdf_cached(TEST_PDF_PATH, { tenantId });
    const initialCallCount = extractor.extractCallCount;
    
    // Generate fingerprint
    const fingerprint = extractor.generateDocumentFingerprint(TEST_PDF_PATH);
    
    // Invalidate the cache
    const invalidated = extractor.invalidateCache(fingerprint, tenantId);
    expect(invalidated).to.be.true;
    
    // Extract again
    await extractor.extract_from_pdf_cached(TEST_PDF_PATH, { tenantId });
    
    // Should have processed again
    expect(extractor.extractCallCount).to.equal(initialCallCount + 1);
  });
  
  it('should provide cache statistics', () => {
    const stats = extractor.getCacheStats();
    
    expect(stats).to.be.an('object');
    expect(stats).to.have.property('totalEntries');
    expect(stats).to.have.property('cacheHits');
    expect(stats).to.have.property('cacheMisses');
    expect(stats).to.have.property('hitRate');
    expect(stats).to.have.property('cacheDirectory');
  });
});

// Run tests if this file is executed directly
if (require.main === module) {
  describe('Document Cache Service Direct Run', function() {
    it('should run tests directly', async () => {
      console.log('Running cache tests...');
      
      // Create extractor
      const extractor = new TestSecurityExtractor({
        debug: true,
        cacheDir: TEST_CACHE_DIR,
        cacheTtl: 60 // 1 minute
      });
      
      // Test fingerprinting
      const fingerprint = extractor.generateDocumentFingerprint(TEST_PDF_PATH);
      console.log(`Document fingerprint: ${fingerprint}`);
      
      // Test caching
      console.log('Testing caching...');
      const result1 = await extractor.extract_from_pdf_cached(TEST_PDF_PATH);
      console.log(`First extraction call count: ${extractor.extractCallCount}`);
      
      const result2 = await extractor.extract_from_pdf_cached(TEST_PDF_PATH);
      console.log(`Second extraction call count: ${extractor.extractCallCount}`);
      
      // Test cache statistics
      const stats = extractor.getCacheStats();
      console.log('Cache stats:', stats);
      
      // Clean up
      if (fs.existsSync(TEST_CACHE_DIR)) {
        fs.rmSync(TEST_CACHE_DIR, { recursive: true, force: true });
      }
      
      console.log('Test completed successfully!');
    });
  });
}

File: DevDocs/backend/tests/test_document_cache.py
"""
Document Cache Service Tests

Tests for the Python implementation of the document caching functionality.
"""

import os
import sys
import unittest
import tempfile
import shutil
import json
import time
from datetime import datetime, timedelta
import hashlib

# Add parent directory to path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import the document cache service
from services.cache.document_cache_service import DocumentCacheService

class TestDocumentCacheService(unittest.TestCase):
    """Test cases for the document cache service."""
    
    def setUp(self):
        """Set up test fixtures."""
        # Create temporary directory for cache
        self.cache_dir = tempfile.mkdtemp()
        
        # Create a test document
        self.test_doc_dir = tempfile.mkdtemp()
        self.test_doc_path = os.path.join(self.test_doc_dir, 'test_document.pdf')
        
        # Create a dummy PDF file
        with open(self.test_doc_path, 'wb') as f:
            f.write(b'%PDF-1.5\nTest PDF content\n%%EOF')
        
        # Create the cache service
        self.cache_service = DocumentCacheService(
            cache_dir=self.cache_dir,
            default_ttl=60,  # 1 minute
            use_tenant_isolation=True
        )
        
        # Test data
        self.test_data = {
            'document_id': 'test123',
            'securities': [
                {'isin': 'US0378331005', 'name': 'Apple Inc.', 'value': 1000},
                {'isin': 'US5949181045', 'name': 'Microsoft Corp.', 'value': 2000}
            ],
            'total_value': 3000,
            'processed_at': datetime.now().isoformat()
        }
    
    def tearDown(self):
        """Tear down test fixtures."""
        # Remove test directories
        shutil.rmtree(self.cache_dir, ignore_errors=True)
        shutil.rmtree(self.test_doc_dir, ignore_errors=True)
    
    def test_generate_document_fingerprint(self):
        """Test document fingerprint generation."""
        # Generate fingerprint
        fingerprint1 = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        fingerprint2 = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        
        # Fingerprints should be consistent
        self.assertEqual(fingerprint1, fingerprint2)
        self.assertTrue(isinstance(fingerprint1, str))
        self.assertTrue(len(fingerprint1) > 0)
        
        # Test with metadata
        metadata = {'document_type': 'portfolio', 'user_id': '12345'}
        fingerprint_with_metadata = self.cache_service.generate_document_fingerprint(
            self.test_doc_path, metadata
        )
        
        # Should be different with metadata
        self.assertNotEqual(fingerprint1, fingerprint_with_metadata)
    
    def test_cache_operations(self):
        """Test basic cache operations."""
        # Generate fingerprint
        fingerprint = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        
        # Initially should not be in cache
        self.assertIsNone(self.cache_service.get_from_cache(fingerprint))
        
        # Save to cache
        self.assertTrue(self.cache_service.save_to_cache(fingerprint, self.test_data))
        
        # Should now be in cache
        cached_data = self.cache_service.get_from_cache(fingerprint)
        self.assertIsNotNone(cached_data)
        self.assertEqual(cached_data['document_id'], self.test_data['document_id'])
        self.assertEqual(len(cached_data['securities']), len(self.test_data['securities']))
        
        # Invalidate cache
        self.assertTrue(self.cache_service.invalidate_cache(fingerprint))
        
        # Should no longer be in cache
        self.assertIsNone(self.cache_service.get_from_cache(fingerprint))
    
    def test_tenant_isolation(self):
        """Test tenant isolation."""
        tenant1 = 'tenant1'
        tenant2 = 'tenant2'
        
        fingerprint = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        
        # Save to cache for tenant1
        self.cache_service.save_to_cache(fingerprint, self.test_data, tenant_id=tenant1)
        
        # Save different data for tenant2
        tenant2_data = self.test_data.copy()
        tenant2_data['document_id'] = 'tenant2-doc'
        self.cache_service.save_to_cache(fingerprint, tenant2_data, tenant_id=tenant2)
        
        # Get from cache for each tenant
        tenant1_cached = self.cache_service.get_from_cache(fingerprint, tenant_id=tenant1)
        tenant2_cached = self.cache_service.get_from_cache(fingerprint, tenant_id=tenant2)
        
        # Should be isolated
        self.assertEqual(tenant1_cached['document_id'], 'test123')
        self.assertEqual(tenant2_cached['document_id'], 'tenant2-doc')
        
        # Tenant directories should exist
        self.assertTrue(os.path.exists(os.path.join(self.cache_dir, tenant1)))
        self.assertTrue(os.path.exists(os.path.join(self.cache_dir, tenant2)))
    
    def test_cache_expiration(self):
        """Test cache expiration."""
        fingerprint = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        
        # Save to cache with 1 second TTL
        self.cache_service.save_to_cache(fingerprint, self.test_data, ttl=1)
        
        # Should be in cache initially
        self.assertIsNotNone(self.cache_service.get_from_cache(fingerprint))
        
        # Wait for expiration
        time.sleep(2)
        
        # Should no longer be in cache
        self.assertIsNone(self.cache_service.get_from_cache(fingerprint))
    
    def test_clear_expired_cache(self):
        """Test clearing expired cache entries."""
        fingerprint1 = 'test-fingerprint-1'
        fingerprint2 = 'test-fingerprint-2'
        
        # Save one with short TTL and one with long TTL
        self.cache_service.save_to_cache(fingerprint1, self.test_data, ttl=1)
        self.cache_service.save_to_cache(fingerprint2, self.test_data, ttl=60)
        
        # Wait for first to expire
        time.sleep(2)
        
        # Clear expired entries
        cleared_count = self.cache_service.clear_expired_cache()
        
        # Should have cleared one entry
        self.assertEqual(cleared_count, 1)
        
        # First should be gone, second should remain
        self.assertIsNone(self.cache_service.get_from_cache(fingerprint1))
        self.assertIsNotNone(self.cache_service.get_from_cache(fingerprint2))
    
    def test_cache_stats(self):
        """Test cache statistics."""
        # Initial stats
        stats = self.cache_service.get_cache_stats()
        self.assertEqual(stats['cache_hits'], 0)
        self.assertEqual(stats['cache_misses'], 0)
        
        fingerprint = self.cache_service.generate_document_fingerprint(self.test_doc_path)
        
        # Miss
        self.cache_service.get_from_cache(fingerprint)
        stats = self.cache_service.get_cache_stats()
        self.assertEqual(stats['cache_misses'], 1)
        
        # Save to cache
        self.cache_service.save_to_cache(fingerprint, self.test_data)
        
        # Hit
        self.cache_service.get_from_cache(fingerprint)
        stats = self.cache_service.get_cache_stats()
        self.assertEqual(stats['cache_hits'], 1)
        self.assertEqual(stats['cache_misses'], 1)


if __name__ == '__main__':
    unittest.main()